/*
 * KLIFS API
 * Dynamically interact with the rich content of KLIFS: the structural kinase database
 *
 * OpenAPI spec version: 0.1
 * 
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package io.swagger.client.model;

import java.util.Objects;
import com.google.gson.annotations.SerializedName;
import io.swagger.annotations.ApiModelProperty;


/**
 * StructureDetails
 */
@javax.annotation.Generated(value = "class io.swagger.codegen.languages.JavaClientCodegen", date = "2016-10-05T09:28:37.207+02:00")
public class StructureDetails   {
  @SerializedName("structure_ID")
  private Integer structureID = null;

  @SerializedName("kinase")
  private String kinase = null;

  @SerializedName("species")
  private String species = null;

  @SerializedName("kinase_ID")
  private Integer kinaseID = null;

  @SerializedName("pdb")
  private String pdb = null;

  @SerializedName("alt")
  private String alt = null;

  @SerializedName("chain")
  private String chain = null;

  @SerializedName("rmsd1")
  private Float rmsd1 = null;

  @SerializedName("rmsd2")
  private Float rmsd2 = null;

  @SerializedName("pocket")
  private String pocket = null;

  @SerializedName("resolution")
  private Float resolution = null;

  @SerializedName("quality_score")
  private Float qualityScore = null;

  @SerializedName("missing_residues")
  private Integer missingResidues = null;

  @SerializedName("missing_atoms")
  private Integer missingAtoms = null;

  @SerializedName("ligand")
  private String ligand = null;

  @SerializedName("allosteric_ligand")
  private String allostericLigand = null;

  @SerializedName("DFG")
  private String DFG = null;

  @SerializedName("aC_helix")
  private String aCHelix = null;

  @SerializedName("Grich_distance")
  private Float grichDistance = null;

  @SerializedName("Grich_angle")
  private Float grichAngle = null;

  @SerializedName("Grich_rotation")
  private Float grichRotation = null;

  @SerializedName("front")
  private Boolean front = null;

  @SerializedName("gate")
  private Boolean gate = null;

  @SerializedName("back")
  private Boolean back = null;

  @SerializedName("fp_I")
  private Boolean fpI = null;

  @SerializedName("fp_II")
  private Boolean fpII = null;

  @SerializedName("bp_I_A")
  private Boolean bpIA = null;

  @SerializedName("bp_I_B")
  private Boolean bpIB = null;

  @SerializedName("bp_II_in")
  private Boolean bpIIIn = null;

  @SerializedName("bp_II_A_in")
  private Boolean bpIIAIn = null;

  @SerializedName("bp_II_B_in")
  private Boolean bpIIBIn = null;

  @SerializedName("bp_II_out")
  private Boolean bpIIOut = null;

  @SerializedName("bp_II_B")
  private Boolean bpIIB = null;

  @SerializedName("bp_III")
  private Boolean bpIII = null;

  @SerializedName("bp_IV")
  private Boolean bpIV = null;

  @SerializedName("bp_V")
  private Boolean bpV = null;

  public StructureDetails structureID(Integer structureID) {
    this.structureID = structureID;
    return this;
  }

   /**
   * KLIFS ID for a specific kinase structure
   * @return structureID
  **/
  @ApiModelProperty(example = "null", required = true, value = "KLIFS ID for a specific kinase structure")
  public Integer getStructureID() {
    return structureID;
  }

  public void setStructureID(Integer structureID) {
    this.structureID = structureID;
  }

  public StructureDetails kinase(String kinase) {
    this.kinase = kinase;
    return this;
  }

   /**
   * HGNC gene symbol for a specific kinase structure
   * @return kinase
  **/
  @ApiModelProperty(example = "null", value = "HGNC gene symbol for a specific kinase structure")
  public String getKinase() {
    return kinase;
  }

  public void setKinase(String kinase) {
    this.kinase = kinase;
  }

  public StructureDetails species(String species) {
    this.species = species;
    return this;
  }

   /**
   * Species of a specific kinase structure (e.g. human)
   * @return species
  **/
  @ApiModelProperty(example = "null", required = true, value = "Species of a specific kinase structure (e.g. human)")
  public String getSpecies() {
    return species;
  }

  public void setSpecies(String species) {
    this.species = species;
  }

  public StructureDetails kinaseID(Integer kinaseID) {
    this.kinaseID = kinaseID;
    return this;
  }

   /**
   * KLIFS ID for a specific kinase
   * @return kinaseID
  **/
  @ApiModelProperty(example = "null", required = true, value = "KLIFS ID for a specific kinase")
  public Integer getKinaseID() {
    return kinaseID;
  }

  public void setKinaseID(Integer kinaseID) {
    this.kinaseID = kinaseID;
  }

  public StructureDetails pdb(String pdb) {
    this.pdb = pdb;
    return this;
  }

   /**
   * 4-letter PDB-code for a given structure
   * @return pdb
  **/
  @ApiModelProperty(example = "null", required = true, value = "4-letter PDB-code for a given structure")
  public String getPdb() {
    return pdb;
  }

  public void setPdb(String pdb) {
    this.pdb = pdb;
  }

  public StructureDetails alt(String alt) {
    this.alt = alt;
    return this;
  }

   /**
   * Indicates the (alternate) model of the given structure
   * @return alt
  **/
  @ApiModelProperty(example = "null", value = "Indicates the (alternate) model of the given structure")
  public String getAlt() {
    return alt;
  }

  public void setAlt(String alt) {
    this.alt = alt;
  }

  public StructureDetails chain(String chain) {
    this.chain = chain;
    return this;
  }

   /**
   * Indicates the chain of the given structure
   * @return chain
  **/
  @ApiModelProperty(example = "null", required = true, value = "Indicates the chain of the given structure")
  public String getChain() {
    return chain;
  }

  public void setChain(String chain) {
    this.chain = chain;
  }

  public StructureDetails rmsd1(Float rmsd1) {
    this.rmsd1 = rmsd1;
    return this;
  }

   /**
   * RMSD compared to the reference template for the superpose pocket
   * @return rmsd1
  **/
  @ApiModelProperty(example = "null", required = true, value = "RMSD compared to the reference template for the superpose pocket")
  public Float getRmsd1() {
    return rmsd1;
  }

  public void setRmsd1(Float rmsd1) {
    this.rmsd1 = rmsd1;
  }

  public StructureDetails rmsd2(Float rmsd2) {
    this.rmsd2 = rmsd2;
    return this;
  }

   /**
   * RMSD compared to the reference template for the full pocket
   * @return rmsd2
  **/
  @ApiModelProperty(example = "null", required = true, value = "RMSD compared to the reference template for the full pocket")
  public Float getRmsd2() {
    return rmsd2;
  }

  public void setRmsd2(Float rmsd2) {
    this.rmsd2 = rmsd2;
  }

  public StructureDetails pocket(String pocket) {
    this.pocket = pocket;
    return this;
  }

   /**
   * Alignment of the 85 pocket residues
   * @return pocket
  **/
  @ApiModelProperty(example = "null", required = true, value = "Alignment of the 85 pocket residues")
  public String getPocket() {
    return pocket;
  }

  public void setPocket(String pocket) {
    this.pocket = pocket;
  }

  public StructureDetails resolution(Float resolution) {
    this.resolution = resolution;
    return this;
  }

   /**
   * Resolution of the crystal struture
   * @return resolution
  **/
  @ApiModelProperty(example = "null", value = "Resolution of the crystal struture")
  public Float getResolution() {
    return resolution;
  }

  public void setResolution(Float resolution) {
    this.resolution = resolution;
  }

  public StructureDetails qualityScore(Float qualityScore) {
    this.qualityScore = qualityScore;
    return this;
  }

   /**
   * The quality score estimates the quality of the structure with respect to the binding pocket as well as the quality of the processing by KLIFS.
   * @return qualityScore
  **/
  @ApiModelProperty(example = "null", required = true, value = "The quality score estimates the quality of the structure with respect to the binding pocket as well as the quality of the processing by KLIFS.")
  public Float getQualityScore() {
    return qualityScore;
  }

  public void setQualityScore(Float qualityScore) {
    this.qualityScore = qualityScore;
  }

  public StructureDetails missingResidues(Integer missingResidues) {
    this.missingResidues = missingResidues;
    return this;
  }

   /**
   * Number of residues missing in the pocket
   * @return missingResidues
  **/
  @ApiModelProperty(example = "null", required = true, value = "Number of residues missing in the pocket")
  public Integer getMissingResidues() {
    return missingResidues;
  }

  public void setMissingResidues(Integer missingResidues) {
    this.missingResidues = missingResidues;
  }

  public StructureDetails missingAtoms(Integer missingAtoms) {
    this.missingAtoms = missingAtoms;
    return this;
  }

   /**
   * Number of atoms missing from pocket residues (not including the missing residues)
   * @return missingAtoms
  **/
  @ApiModelProperty(example = "null", required = true, value = "Number of atoms missing from pocket residues (not including the missing residues)")
  public Integer getMissingAtoms() {
    return missingAtoms;
  }

  public void setMissingAtoms(Integer missingAtoms) {
    this.missingAtoms = missingAtoms;
  }

  public StructureDetails ligand(String ligand) {
    this.ligand = ligand;
    return this;
  }

   /**
   * 3-letter PDB-code of the ligand within the main pocket
   * @return ligand
  **/
  @ApiModelProperty(example = "null", value = "3-letter PDB-code of the ligand within the main pocket")
  public String getLigand() {
    return ligand;
  }

  public void setLigand(String ligand) {
    this.ligand = ligand;
  }

  public StructureDetails allostericLigand(String allostericLigand) {
    this.allostericLigand = allostericLigand;
    return this;
  }

   /**
   * 3-letter PDB-code of the ligand outside the main pocket
   * @return allostericLigand
  **/
  @ApiModelProperty(example = "null", value = "3-letter PDB-code of the ligand outside the main pocket")
  public String getAllostericLigand() {
    return allostericLigand;
  }

  public void setAllostericLigand(String allostericLigand) {
    this.allostericLigand = allostericLigand;
  }

  public StructureDetails DFG(String DFG) {
    this.DFG = DFG;
    return this;
  }

   /**
   * Conformation of the DFG motif
   * @return DFG
  **/
  @ApiModelProperty(example = "null", value = "Conformation of the DFG motif")
  public String getDFG() {
    return DFG;
  }

  public void setDFG(String DFG) {
    this.DFG = DFG;
  }

  public StructureDetails aCHelix(String aCHelix) {
    this.aCHelix = aCHelix;
    return this;
  }

   /**
   * Conformation of the alpha-C helix
   * @return aCHelix
  **/
  @ApiModelProperty(example = "null", value = "Conformation of the alpha-C helix")
  public String getACHelix() {
    return aCHelix;
  }

  public void setACHelix(String aCHelix) {
    this.aCHelix = aCHelix;
  }

  public StructureDetails grichDistance(Float grichDistance) {
    this.grichDistance = grichDistance;
    return this;
  }

   /**
   * Conformation of the G-rich loop - distance
   * @return grichDistance
  **/
  @ApiModelProperty(example = "null", value = "Conformation of the G-rich loop - distance")
  public Float getGrichDistance() {
    return grichDistance;
  }

  public void setGrichDistance(Float grichDistance) {
    this.grichDistance = grichDistance;
  }

  public StructureDetails grichAngle(Float grichAngle) {
    this.grichAngle = grichAngle;
    return this;
  }

   /**
   * Conformation of the G-rich loop - angle
   * @return grichAngle
  **/
  @ApiModelProperty(example = "null", value = "Conformation of the G-rich loop - angle")
  public Float getGrichAngle() {
    return grichAngle;
  }

  public void setGrichAngle(Float grichAngle) {
    this.grichAngle = grichAngle;
  }

  public StructureDetails grichRotation(Float grichRotation) {
    this.grichRotation = grichRotation;
    return this;
  }

   /**
   * Conformation of the G-rich loop - rotation
   * @return grichRotation
  **/
  @ApiModelProperty(example = "null", value = "Conformation of the G-rich loop - rotation")
  public Float getGrichRotation() {
    return grichRotation;
  }

  public void setGrichRotation(Float grichRotation) {
    this.grichRotation = grichRotation;
  }

  public StructureDetails front(Boolean front) {
    this.front = front;
    return this;
  }

   /**
   * The ligand binds to the front pocket
   * @return front
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to the front pocket")
  public Boolean getFront() {
    return front;
  }

  public void setFront(Boolean front) {
    this.front = front;
  }

  public StructureDetails gate(Boolean gate) {
    this.gate = gate;
    return this;
  }

   /**
   * The ligand binds to the gate area
   * @return gate
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to the gate area")
  public Boolean getGate() {
    return gate;
  }

  public void setGate(Boolean gate) {
    this.gate = gate;
  }

  public StructureDetails back(Boolean back) {
    this.back = back;
    return this;
  }

   /**
   * The ligand binds to the back pocket
   * @return back
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to the back pocket")
  public Boolean getBack() {
    return back;
  }

  public void setBack(Boolean back) {
    this.back = back;
  }

  public StructureDetails fpI(Boolean fpI) {
    this.fpI = fpI;
    return this;
  }

   /**
   * The ligand binds to FP-I
   * @return fpI
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to FP-I")
  public Boolean getFpI() {
    return fpI;
  }

  public void setFpI(Boolean fpI) {
    this.fpI = fpI;
  }

  public StructureDetails fpII(Boolean fpII) {
    this.fpII = fpII;
    return this;
  }

   /**
   * The ligand binds to FP-II
   * @return fpII
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to FP-II")
  public Boolean getFpII() {
    return fpII;
  }

  public void setFpII(Boolean fpII) {
    this.fpII = fpII;
  }

  public StructureDetails bpIA(Boolean bpIA) {
    this.bpIA = bpIA;
    return this;
  }

   /**
   * The ligand binds to BP-I-A
   * @return bpIA
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-I-A")
  public Boolean getBpIA() {
    return bpIA;
  }

  public void setBpIA(Boolean bpIA) {
    this.bpIA = bpIA;
  }

  public StructureDetails bpIB(Boolean bpIB) {
    this.bpIB = bpIB;
    return this;
  }

   /**
   * The ligand binds to BP-I-B
   * @return bpIB
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-I-B")
  public Boolean getBpIB() {
    return bpIB;
  }

  public void setBpIB(Boolean bpIB) {
    this.bpIB = bpIB;
  }

  public StructureDetails bpIIIn(Boolean bpIIIn) {
    this.bpIIIn = bpIIIn;
    return this;
  }

   /**
   * The ligand binds to BP-II-in
   * @return bpIIIn
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-II-in")
  public Boolean getBpIIIn() {
    return bpIIIn;
  }

  public void setBpIIIn(Boolean bpIIIn) {
    this.bpIIIn = bpIIIn;
  }

  public StructureDetails bpIIAIn(Boolean bpIIAIn) {
    this.bpIIAIn = bpIIAIn;
    return this;
  }

   /**
   * The ligand binds to BP-II-A-in
   * @return bpIIAIn
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-II-A-in")
  public Boolean getBpIIAIn() {
    return bpIIAIn;
  }

  public void setBpIIAIn(Boolean bpIIAIn) {
    this.bpIIAIn = bpIIAIn;
  }

  public StructureDetails bpIIBIn(Boolean bpIIBIn) {
    this.bpIIBIn = bpIIBIn;
    return this;
  }

   /**
   * The ligand binds to BP-II-B-in
   * @return bpIIBIn
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-II-B-in")
  public Boolean getBpIIBIn() {
    return bpIIBIn;
  }

  public void setBpIIBIn(Boolean bpIIBIn) {
    this.bpIIBIn = bpIIBIn;
  }

  public StructureDetails bpIIOut(Boolean bpIIOut) {
    this.bpIIOut = bpIIOut;
    return this;
  }

   /**
   * The ligand binds to BP-II-out
   * @return bpIIOut
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-II-out")
  public Boolean getBpIIOut() {
    return bpIIOut;
  }

  public void setBpIIOut(Boolean bpIIOut) {
    this.bpIIOut = bpIIOut;
  }

  public StructureDetails bpIIB(Boolean bpIIB) {
    this.bpIIB = bpIIB;
    return this;
  }

   /**
   * The ligand binds to BP-II-B
   * @return bpIIB
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-II-B")
  public Boolean getBpIIB() {
    return bpIIB;
  }

  public void setBpIIB(Boolean bpIIB) {
    this.bpIIB = bpIIB;
  }

  public StructureDetails bpIII(Boolean bpIII) {
    this.bpIII = bpIII;
    return this;
  }

   /**
   * The ligand binds to BP-III
   * @return bpIII
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-III")
  public Boolean getBpIII() {
    return bpIII;
  }

  public void setBpIII(Boolean bpIII) {
    this.bpIII = bpIII;
  }

  public StructureDetails bpIV(Boolean bpIV) {
    this.bpIV = bpIV;
    return this;
  }

   /**
   * The ligand binds to BP-IV
   * @return bpIV
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-IV")
  public Boolean getBpIV() {
    return bpIV;
  }

  public void setBpIV(Boolean bpIV) {
    this.bpIV = bpIV;
  }

  public StructureDetails bpV(Boolean bpV) {
    this.bpV = bpV;
    return this;
  }

   /**
   * The ligand binds to BP-V
   * @return bpV
  **/
  @ApiModelProperty(example = "null", value = "The ligand binds to BP-V")
  public Boolean getBpV() {
    return bpV;
  }

  public void setBpV(Boolean bpV) {
    this.bpV = bpV;
  }


  @Override
  public boolean equals(java.lang.Object o) {
    if (this == o) {
      return true;
    } else if (o == null || getClass() != o.getClass()) {
      return false;
    } else {
      StructureDetails structureDetails = (StructureDetails) o;
      return Objects.equals(this.structureID, structureDetails.structureID) &&
        Objects.equals(this.kinase, structureDetails.kinase) &&
        Objects.equals(this.species, structureDetails.species) &&
        Objects.equals(this.kinaseID, structureDetails.kinaseID) &&
        Objects.equals(this.pdb, structureDetails.pdb) &&
        Objects.equals(this.alt, structureDetails.alt) &&
        Objects.equals(this.chain, structureDetails.chain) &&
        Objects.equals(this.rmsd1, structureDetails.rmsd1) &&
        Objects.equals(this.rmsd2, structureDetails.rmsd2) &&
        Objects.equals(this.pocket, structureDetails.pocket) &&
        Objects.equals(this.resolution, structureDetails.resolution) &&
        Objects.equals(this.qualityScore, structureDetails.qualityScore) &&
        Objects.equals(this.missingResidues, structureDetails.missingResidues) &&
        Objects.equals(this.missingAtoms, structureDetails.missingAtoms) &&
        Objects.equals(this.ligand, structureDetails.ligand) &&
        Objects.equals(this.allostericLigand, structureDetails.allostericLigand) &&
        Objects.equals(this.DFG, structureDetails.DFG) &&
        Objects.equals(this.aCHelix, structureDetails.aCHelix) &&
        Objects.equals(this.grichDistance, structureDetails.grichDistance) &&
        Objects.equals(this.grichAngle, structureDetails.grichAngle) &&
        Objects.equals(this.grichRotation, structureDetails.grichRotation) &&
        Objects.equals(this.front, structureDetails.front) &&
        Objects.equals(this.gate, structureDetails.gate) &&
        Objects.equals(this.back, structureDetails.back) &&
        Objects.equals(this.fpI, structureDetails.fpI) &&
        Objects.equals(this.fpII, structureDetails.fpII) &&
        Objects.equals(this.bpIA, structureDetails.bpIA) &&
        Objects.equals(this.bpIB, structureDetails.bpIB) &&
        Objects.equals(this.bpIIIn, structureDetails.bpIIIn) &&
        Objects.equals(this.bpIIAIn, structureDetails.bpIIAIn) &&
        Objects.equals(this.bpIIBIn, structureDetails.bpIIBIn) &&
        Objects.equals(this.bpIIOut, structureDetails.bpIIOut) &&
        Objects.equals(this.bpIIB, structureDetails.bpIIB) &&
        Objects.equals(this.bpIII, structureDetails.bpIII) &&
        Objects.equals(this.bpIV, structureDetails.bpIV) &&
        Objects.equals(this.bpV, structureDetails.bpV);
     }
  }

  @Override
  public int hashCode() {
    return Objects.hash(structureID, kinase, species, kinaseID, pdb, alt, chain, rmsd1, rmsd2, pocket, resolution, qualityScore, missingResidues, missingAtoms, ligand, allostericLigand, DFG, aCHelix, grichDistance, grichAngle, grichRotation, front, gate, back, fpI, fpII, bpIA, bpIB, bpIIIn, bpIIAIn, bpIIBIn, bpIIOut, bpIIB, bpIII, bpIV, bpV);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class StructureDetails {\n");
    
    sb.append("    structureID: ").append(toIndentedString(structureID)).append("\n");
    sb.append("    kinase: ").append(toIndentedString(kinase)).append("\n");
    sb.append("    species: ").append(toIndentedString(species)).append("\n");
    sb.append("    kinaseID: ").append(toIndentedString(kinaseID)).append("\n");
    sb.append("    pdb: ").append(toIndentedString(pdb)).append("\n");
    sb.append("    alt: ").append(toIndentedString(alt)).append("\n");
    sb.append("    chain: ").append(toIndentedString(chain)).append("\n");
    sb.append("    rmsd1: ").append(toIndentedString(rmsd1)).append("\n");
    sb.append("    rmsd2: ").append(toIndentedString(rmsd2)).append("\n");
    sb.append("    pocket: ").append(toIndentedString(pocket)).append("\n");
    sb.append("    resolution: ").append(toIndentedString(resolution)).append("\n");
    sb.append("    qualityScore: ").append(toIndentedString(qualityScore)).append("\n");
    sb.append("    missingResidues: ").append(toIndentedString(missingResidues)).append("\n");
    sb.append("    missingAtoms: ").append(toIndentedString(missingAtoms)).append("\n");
    sb.append("    ligand: ").append(toIndentedString(ligand)).append("\n");
    sb.append("    allostericLigand: ").append(toIndentedString(allostericLigand)).append("\n");
    sb.append("    DFG: ").append(toIndentedString(DFG)).append("\n");
    sb.append("    aCHelix: ").append(toIndentedString(aCHelix)).append("\n");
    sb.append("    grichDistance: ").append(toIndentedString(grichDistance)).append("\n");
    sb.append("    grichAngle: ").append(toIndentedString(grichAngle)).append("\n");
    sb.append("    grichRotation: ").append(toIndentedString(grichRotation)).append("\n");
    sb.append("    front: ").append(toIndentedString(front)).append("\n");
    sb.append("    gate: ").append(toIndentedString(gate)).append("\n");
    sb.append("    back: ").append(toIndentedString(back)).append("\n");
    sb.append("    fpI: ").append(toIndentedString(fpI)).append("\n");
    sb.append("    fpII: ").append(toIndentedString(fpII)).append("\n");
    sb.append("    bpIA: ").append(toIndentedString(bpIA)).append("\n");
    sb.append("    bpIB: ").append(toIndentedString(bpIB)).append("\n");
    sb.append("    bpIIIn: ").append(toIndentedString(bpIIIn)).append("\n");
    sb.append("    bpIIAIn: ").append(toIndentedString(bpIIAIn)).append("\n");
    sb.append("    bpIIBIn: ").append(toIndentedString(bpIIBIn)).append("\n");
    sb.append("    bpIIOut: ").append(toIndentedString(bpIIOut)).append("\n");
    sb.append("    bpIIB: ").append(toIndentedString(bpIIB)).append("\n");
    sb.append("    bpIII: ").append(toIndentedString(bpIII)).append("\n");
    sb.append("    bpIV: ").append(toIndentedString(bpIV)).append("\n");
    sb.append("    bpV: ").append(toIndentedString(bpV)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(java.lang.Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}

